src/common/offline_transformations/src/pruning/init_masks.cpp

```diff
@@ -75,8 +76,34 @@ class ngraph::pass::init_masks::InitMatMulMask : public MatcherPass {
             NodeVector weights_calculation_nodes;
             auto cur_node = matmul->get_input_node_shared_ptr(1);
 
+            if (!cur_node->output(0).get_partial_shape().is_static()) return false;
+            auto dim_order = std::vector<int64_t>();
+            const auto input_size = cur_node->get_shape().size();
+            for (int64_t idx = 0; idx < input_size; ++idx)
+                dim_order.push_back(idx);
+
             while (!ngraph::is_type<opset6::Constant>(cur_node) && cur_node->inputs().size()) {
                 weights_calculation_nodes.push_back(cur_node);
+                if (ngraph::is_type<opset6::Transpose>(cur_node)) {
+                    const auto forward_order = get_constant_from_source(cur_node->get_input_node_shared_ptr(1));
+                    if (!forward_order) return false;
+                    const auto forward_order_vec = forward_order->cast_vector<int64_t>();
+                    if (forward_order_vec.size() != input_size) return false;
+                    auto new_order = std::vector<int64_t>();
+                    for (size_t i = 0; i < forward_order_vec.size(); ++i) {
+                        const auto dim = std::find(forward_order_vec.begin(), forward_order_vec.end(), i) - forward_order_vec.begin();
+                        // Dim should be valid because of transpose operation input_order input restrictions
+                        new_order.push_back(dim_order[dim]);
+                    }
+                    dim_order = new_order;
+                } else {
+                    if (ngraph::is_type<opset6::Reshape>(cur_node) || ngraph::is_type<opset6::MatMul>(cur_node)) {
+                        NGRAPH_DEBUG << "Can't init mask for MatMul: " <<
+                        matmul->get_friendly_name() << " because of node " <<
+                        cur_node->get_friendly_name() << " in the way from weights to Matmul" << std::endl;
+                        return false;
+                    }
+                }
```

Comment:
do I understand correctly that this is for matmul weights? what if we have both reshape and transpose hanging there?

Reply:
This code is trying to find matmul column axis in its "right" constant (second input). Reshape ops in a path from matmul to its weights isn't supported by now, so we aren't pruning columns in such matmul. So, it should work ok in case both reshape and transpose hanging in path (tested [here](https://github.com/openvinotoolkit/openvino/blob/1eaed0819ee8aeeea2691cd9f00e0668942791c0/src/tests/functional/inference_engine/transformations/offline_transformations/pruning_test.cpp#L3688-L3727))

Reply:
as I understand this, we try to find Const, but if this ops chain (highlighted) contains Reshape or Matmul ops, we do not support such cases, right?
![image](https://user-images.githubusercontent.com/21181928/181216052-714521ca-9d8d-4916-9063-4077a6d0e441.png)

Reply:
@itikhono yes, you are right. We are skipping const search in case any operation which change shapes are found. We now considering only transpose and reshape operations, but other operation (like conv or split or other exotic patterns) aren't supported